### PR TITLE
swap out rust-crypto for RustCrypto libraries for key management

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

#1245

## Proposed Changes

Swap out `rust-crypto` with libraries from here: https://github.com/RustCrypto
in the  `eth2_key_derivation` and `eth2_keystore`

## Additional Info

The [stream-ciphers](https://github.com/RustCrypto/stream-ciphers) library had this warning:

> Crates have not yet received any formal cryptographic and security reviews.

### CODE REVIEW

**Comment by paulhauner** [comment]
Thanks for this @realbigsean! I'm keen to see this get merged, it's on my todo list :)

File: crypto/eth2_keystore/src/keystore.rs [hunk 1]
```diff
@@ -280,15 +286,13 @@ pub fn encrypt(
...
+    let mut cipher_text = plain_text.to_vec();
     match &cipher {
         Cipher::Aes128Ctr(params) => {
-            crypto::aes::ctr(
-                crypto::aes::KeySize::KeySize128,
-                &derived_key.as_bytes()[0..16],
-                params.iv.as_bytes(),
-            )
-            .process(plain_text, &mut cipher_text);
+            let key = GenericArray::from_slice(&derived_key.as_bytes()[0..16]);
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply crypto/eth2_keystore/src/keystore.rs:292]: Oh wait, looks like I didn't understand how `GenericArray` works! I thought it owned the array, but it does some [`unsafe` stuff](https://github.com/fizyk20/generic-array/blob/1a55462c109f18abeb0372f421ecee3d81d24658/src/lib.rs#L623-L630) to transmute the slice into a `GenericArray`.

Please ignore this comment :)

File: crypto/eth2_key_derivation/src/path.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 use crate::{lamport_secret_key::LamportSecretKey, plain_text::PlainText};
```
[the remark this answers is not in the document]
Comment (realbigsean) [reply crypto/eth2_key_derivation/src/path.rs:1]: Ah ok, no worries

File: crypto/eth2_keystore/src/keystore.rs [hunk 3]
```diff
@@ -316,21 +320,20 @@ pub fn decrypt(password: &[u8], crypto: &Crypto) -> Result<PlainText, Error> {
...
+            // NOTE: we do not check the size of the `iv` as there is no guidance about
+            // this on EIP-2335.
+            //
+            // Reference:
+            //
+            // - https://github.com/ethereum/EIPs/issues/2339#issuecomment-623865023
+            let key = GenericArray::from_slice(&derived_key.as_bytes()[0..16]);
+            let nonce = GenericArray::from_slice(params.iv.as_bytes());
+            let mut cipher = AesCtr::new(&key, &nonce);
+            cipher.seek(0);
```
[the remark this answers is not in the document]
Comment (realbigsean) [reply crypto/eth2_keystore/src/keystore.rs:335]: Yep, think it's redundant. Based on the example in the docs, I thought you needed it for decryption, but I think it's only necessary if you use `apply_keystream()` on the same cipher instance more than once.

File: crypto/eth2_keystore/src/keystore.rs [hunk 4]
```diff
@@ -400,12 +401,14 @@ fn derive_key(password: &[u8], kdf: &Kdf) -> Result<DerivedKey, Error> {
                 return Err(Error::InvalidScryptParam);
             }
 
-            crypto::scrypt::scrypt(
+            scrypt(
                 password,
                 params.salt.as_bytes(),
-                &crypto::scrypt::ScryptParams::new(log2_int(params.n) as u8, params.r, params.p),
+                &ScryptParams::new(log2_int(params.n) as u8, params.r, params.p)
+                    .expect("Invalid params for scrypt private key decryption."),
```
[the remark this answers is not in the document]
Comment (realbigsean) [reply crypto/eth2_keystore/src/keystore.rs]: Updated this, let me know if it looks off. Not too familiar with error handling in Rust

**Comment by paulhauner** [comment]
Thanks for the changes @realbigsean! We merged our `spec-v0.12` branch into `master` yesterday, looks like we have some conflicts sorry.

**Comment by realbigsean** [comment]
> Thanks for the changes @realbigsean! We merged our `spec-v0.12` branch into `master` yesterday, looks like we have some conflicts sorry.

No problem, just resolved the conflicts.

**Pull request merged**